pkg/ccl/backupccl/restore_job.go

```diff
@@ -2010,51 +2018,98 @@ func insertStats(
 		return nil
 	}
 
-	totalNumBatches := len(latestStats) / restoreStatsInsertBatchSize
-	log.Infof(ctx, "restore will insert %d TableStatistics in %d batches", len(latestStats), totalNumBatches)
-	insertStatsProgress := log.Every(10 * time.Second)
-
 	// We could be restoring hundreds of tables, so insert the new stats in
 	// batches instead of all in a single, long-running txn. This prevents intent
 	// buildup in the face of txn retries.
-	for {
-		if len(latestStats) == 0 {
-			return nil
-		}
-
-		if len(latestStats) < restoreStatsInsertBatchSize {
-			restoreStatsInsertBatchSize = len(latestStats)
-		}
-
-		if err := execCfg.InternalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
-			if err := stats.InsertNewStats(
-				ctx, execCfg.Settings, txn, latestStats[:restoreStatsInsertBatchSize],
-			); err != nil {
-				return errors.Wrapf(err, "inserting stats from backup")
-			}
+	batchSize := restoreStatsInsertBatchSize
+	totalNumBatches := len(latestStats) / batchSize
+	if len(latestStats)%batchSize != 0 {
+		totalNumBatches += 1
+	}
+	log.Infof(ctx, "restore will insert %d TableStatistics in %d batches", len(latestStats), totalNumBatches)
+	insertStatsProgress := log.Every(10 * time.Second)
 
-			// If this is the last batch, mark the stats insertion complete.
-			if restoreStatsInsertBatchSize == len(latestStats) {
-				details.StatsInserted = true
-				if err := job.WithTxn(txn).SetDetails(ctx, details); err != nil {
-					return errors.Wrapf(err, "updating job marking stats insertion complete")
-				}
+	startingStatsInsertion := timeutil.Now()
+	// Should never block.
+	batchCh := make(chan []*stats.TableStatisticProto, totalNumBatches)
+	generateStatsBatch := func(ctx context.Context) error {
+		defer close(batchCh)
+		for {
+			if len(latestStats) == 0 {
 				return nil
 			}
+			if len(latestStats) < batchSize {
+				batchSize = len(latestStats)
+			}
+			batchCh <- latestStats[:batchSize]
 
-			return nil
-		}); err != nil {
-			return err
+			// Truncate the stats that we have inserted in the txn above.
+			latestStats = latestStats[batchSize:]
 		}
+	}
 
-		// Truncate the stats that we have inserted in the txn above.
-		latestStats = latestStats[restoreStatsInsertBatchSize:]
-
-		if insertStatsProgress.ShouldLog() {
-			remainingBatches := len(latestStats) / restoreStatsInsertBatchSize
-			log.Infof(ctx, "restore has %d/%d TableStatistics batches remaining to insert", remainingBatches, totalNumBatches)
+	logStatsProgress := func(remainingBatches, completedBatches int) {
+		msg := fmt.Sprintf("restore has %d/%d TableStatistics batches remaining to insert",
+			remainingBatches, totalNumBatches)
+		var rate int
```

Comment:
[nit] This is only used in the if block below, I don't think we need to declare it here.